### PR TITLE
caddyfile: reject cyclic imports

### PR DESCRIPTION
**Pull request by Mohammed90**
Fix an edge case of Caddyfile import itself. I don't imagine anyone deliberately self-importing a Caddyfile but could happen if glob expansion wasn't vetted by the user.

### CODE REVIEW

**Comment by Mohammed90** [comment]
Converting it to draft as @francislavoie and I found more corner cases

**Comment by Mohammed90** [comment]
I opted for minimal implementation of the graph to avoid pulling an additional dep just for tracking cycles.

File: caddyconfig/caddyfile/lexer.go [hunk 1]
```diff
@@ -35,9 +35,11 @@ type (
...
 	// Token represents a single parsable unit.
 	Token struct {
-		File string
-		Line int
-		Text string
+		File        string
+		Line        int
+		Text        string
+		InSnippet   bool
+		SnippetName string
```
Comment (mholt) [inline caddyconfig/caddyfile/lexer.go:42]: It looks like these are used only in this package, right? Let's unexport them.

File: caddyconfig/caddyfile/parse.go [hunk 2]
```diff
@@ -165,6 +173,12 @@ func (p *parser) begin() error {
 		if err != nil {
 			return err
 		}
+		// Keep track of which tokens come from snippets and what the snippet name is
```
Comment (mholt) [inline caddyconfig/caddyfile/parse.go:176]: Maybe this comment should explain why; right now it doesn't add too much to what is already obvious from the lines below.

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks for contributing this. Might save some people some head scratching.

I have just a couple requests, nothing major.

File: caddyconfig/caddyfile/lexer.go (hunk 1 above)
Comment (Mohammed90) [reply caddyconfig/caddyfile/lexer.go]: Done

File: caddyconfig/caddyfile/parse.go (hunk 2 above)
Comment (Mohammed90) [reply caddyconfig/caddyfile/parse.go]: Done

**Review by mholt — APPROVED** [review]
Cool, hopefully this will help a few people and prevent some problems! Thanks for contributing this @Mohammed90 !

**Pull request merged**